contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol

```diff
@@ -0,0 +1,1226 @@
+import {Address} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/Address.sol";
+import {StreamsLookupCompatibleInterface} from "../interfaces/StreamsLookupCompatibleInterface.sol";
+import {ILogAutomation, Log} from "../interfaces/ILogAutomation.sol";
+import {IZKSyncAutomationForwarder} from "../interfaces/zksync/IZKSyncAutomationForwarder.sol";
+import {ConfirmedOwner} from "../../shared/access/ConfirmedOwner.sol";
+import {AggregatorV3Interface} from "../../shared/interfaces/AggregatorV3Interface.sol";
+import {LinkTokenInterface} from "../../shared/interfaces/LinkTokenInterface.sol";
+import {KeeperCompatibleInterface} from "../interfaces/KeeperCompatibleInterface.sol";
+import {IChainModule} from "../interfaces/IChainModule.sol";
+import {IERC20Metadata as IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {SafeCast} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/math/SafeCast.sol";
+import {IWrappedNative} from "../interfaces/v2_3/IWrappedNative.sol";
+
+/**
+ * @notice Base Keeper Registry contract, contains shared logic between
+ * AutomationRegistry and AutomationRegistryLogic
+ * @dev all errors, events, and internal functions should live here
+ */
+// solhint-disable-next-line max-states-count
+abstract contract ZKSyncAutomationRegistryBase2_3 is ConfirmedOwner {
+  using Address for address;
+  using EnumerableSet for EnumerableSet.UintSet;
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  address internal constant ZERO_ADDRESS = address(0);
+  address internal constant IGNORE_ADDRESS = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
+  bytes4 internal constant CHECK_SELECTOR = KeeperCompatibleInterface.checkUpkeep.selector;
+  bytes4 internal constant PERFORM_SELECTOR = KeeperCompatibleInterface.performUpkeep.selector;
+  bytes4 internal constant CHECK_CALLBACK_SELECTOR = StreamsLookupCompatibleInterface.checkCallback.selector;
+  bytes4 internal constant CHECK_LOG_SELECTOR = ILogAutomation.checkLog.selector;
+  uint256 internal constant PERFORM_GAS_MIN = 2_300;
+  uint256 internal constant CANCELLATION_DELAY = 50;
+  uint256 internal constant PERFORM_GAS_CUSHION = 5_000;
+  uint256 internal constant PPB_BASE = 1_000_000_000;
+  uint32 internal constant UINT32_MAX = type(uint32).max;
+  // The first byte of the mask can be 0, because we only ever have 31 oracles
+  uint256 internal constant ORACLE_MASK = 0x0001010101010101010101010101010101010101010101010101010101010101;
+  uint8 internal constant UPKEEP_VERSION_BASE = 4;
+
+  // Next block of constants are only used in maxPayment estimation during checkUpkeep simulation
+  // These values are calibrated using hardhat tests which simulate various cases and verify that
+  // the variables result in accurate estimation
+  uint256 internal constant REGISTRY_CONDITIONAL_OVERHEAD = 92_000; // Fixed gas overhead for conditional upkeeps
```

Comment:
conditional and log trigger overheads are reduced slightly due to simpler logic in hot path perform